### PR TITLE
PMM-6633 Flexible PMM Client Docker image

### PR DESCRIPTION
**Pull request by AlekSi**

### CODE REVIEW

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-f2f8db5
client docker - perconalab/pmm-client-fb:PR-1184-f2f8db5
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-f2f8db5.tar.gz

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-c05af1f
client docker - perconalab/pmm-client-fb:PR-1184-c05af1f
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-c05af1f.tar.gz

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4623/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4005/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5211/

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4625/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4007/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5213/

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-42a5074
client docker - perconalab/pmm-client-fb:PR-1184-42a5074
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-42a5074.tar.gz

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4627/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4009/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5215/

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-a990424
client docker - perconalab/pmm-client-fb:PR-1184-a990424
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-a990424.tar.gz

File: build/pmm-client-docker/entrypoint.py [hunk 1]
```diff
@@ -41,15 +44,39 @@ def main():
...
             sys.exit(status)
 
-    print('Starting pmm-agent ...', file=sys.stderr)
-    sys.stderr.flush()
+    if PMM_AGENT_PRERUN_SCRIPT:
+        print('Starting pmm-agent for prerun script ...', file=sys.stderr)
+        agent = subprocess.Popen(['pmm-agent', 'run'])
+
+        print('Running {} ...'.format(PMM_AGENT_PRERUN_SCRIPT), file=sys.stderr)
+        status = subprocess.call([PMM_AGENT_PRERUN_SCRIPT])
```
Comment (delgod) [inline build/pmm-client-docker/entrypoint.py:57]: there are two requirements
1) command like `sh -c "ls;id"` should work fine, just test with `PMM_AGENT_PRERUN_SCRIPT = 'sh -c "ls;id"'`
2) we should see stdout and stderr of command on the main stderr

**Review by delgod — CHANGES_REQUESTED** [review]

File: build/pmm-client-docker/entrypoint.py (hunk 1 above)
Comment (AlekSi) [reply build/pmm-client-docker/entrypoint.py:57]: `PMM_AGENT_PRERUN_SCRIPT ` contains file name that should be added via volume or some other means, not the script itself.

stdout and stderr should already be visible. Why you want both stdout and stderr in stderr?

Comment (delgod) [reply build/pmm-client-docker/entrypoint.py:57]: > contains file name that should be added via volume
it is not the right thing, please let's just add one call
```suggestion
        status = subprocess.call(shlex.split(PMM_AGENT_PRERUN_SCRIPT))
```

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-f3a8e9c
client docker - perconalab/pmm-client-fb:PR-1184-f3a8e9c
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-f3a8e9c.tar.gz

File: build/pmm-client-docker/entrypoint.py (hunk 1 above)
Comment (AlekSi) [reply build/pmm-client-docker/entrypoint.py]:
```
export PMM_AGENT_PRERUN_SCRIPT='ls -la | grep Dockerfile && echo ls yes || echo ls no; id -u | grep 1000 && echo id yes || echo id no'
./entrypoint.py
```

now produces

```
Running prerun shell script ...
-rw-r--r--  1 aleksi  staff   728 Jun  9 15:48 Dockerfile
ls yes
id no
Prerun shell script exited with 0.
```

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4630/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4012/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5218/

File: build/pmm-client-docker/entrypoint.py [hunk 2]
```diff
@@ -41,15 +46,45 @@ def main():
...
+                break
+            time.sleep(1)
+        if agent.returncode is None:
+            print('Killing pmm-agent ...', file=sys.stderr)
+            agent.kill()
+
+        agent.wait()
+
+        if status != 0:
+            sys.exit(status)
```
Comment (delgod) [inline build/pmm-client-docker/entrypoint.py:83]: if both commands passed only status of the last command counted here

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4631/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4013/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5219/

File: build/pmm-client-docker/entrypoint.py [hunk 3]
```diff
@@ -32,7 +37,11 @@ def main():
 
     if len(sys.argv) > 1:
         print(__doc__, file=sys.stderr)
-        os.system('pmm-agent setup --help')
+        subprocess.call(['pmm-agent', 'setup', '--help'])
+        sys.exit(1)
+
+    if PMM_AGENT_PRERUN_FILE and PMM_AGENT_PRERUN_SCRIPT:
+        print('Both PMM_AGENT_PRERUN_FILE and PMM_AGENT_PRERUN_SCRIPT cannot be set.', file=sys.stderr)
         sys.exit(1)
```
Comment (AlekSi) [inline build/pmm-client-docker/entrypoint.py:45]: @delgod ^

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-8b9184f
client docker - perconalab/pmm-client-fb:PR-1184-8b9184f
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-8b9184f.tar.gz

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-18f09bb
client docker - perconalab/pmm-client-fb:PR-1184-18f09bb
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-18f09bb.tar.gz

**Review by BupycHuk — APPROVED** [review]

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-bb256aa
client docker - perconalab/pmm-client-fb:PR-1184-bb256aa
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-bb256aa.tar.gz

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-bb256aa
client docker - perconalab/pmm-client-fb:PR-1184-bb256aa
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-bb256aa.tar.gz

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-bb256aa
client docker - perconalab/pmm-client-fb:PR-1184-bb256aa
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-bb256aa.tar.gz

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4632/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4014/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5220/

**Review by askomorokhov — APPROVED** [review]

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4633/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4015/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5221/

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4634/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4016/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5222/

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4635/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4017/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5223/

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4636/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4018/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5224/

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-1184-bb256aa
client docker - perconalab/pmm-client-fb:PR-1184-bb256aa
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1184-bb256aa.tar.gz

**Comment by JNKPercona** [comment]
Some Tests have Failed Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/4639/ BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/4021/ & UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/5227/

**Pull request merged**